pkg/dao/crd/dao.go

```diff
@@ -0,0 +1,306 @@
+		return nil, err
+	}
+	return convertServiceInstanceToAPB(servInstance.Spec, spec, servInstance.GetName()), nil
+}
+
+// SetServiceInstance - Set service instance for an id in the kvp API.
+func (d *Dao) SetServiceInstance(id string, serviceInstance *apb.ServiceInstance) error {
+	log.Debugf("set service instance: %v", id)
+	spec := convertServiceInstanceToCRD(serviceInstance)
+	if si, err := d.client.ServiceInstances(d.namespace).Get(id, metav1.GetOptions{}); err == nil {
+		log.Debugf("updateing service instance: %v", id)
+		si.Spec = spec
+		_, err := d.client.ServiceInstances(d.namespace).Update(si)
+		if err != nil {
+			log.Errorf("unable to update service instance - %v", err)
+			return err
+		}
+		return nil
+	}
+	s := v1.ServiceInstance{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      id,
+			Namespace: d.namespace,
+		},
+		Spec: spec,
+	}
+
+	_, err := d.client.ServiceInstances(d.namespace).Create(&s)
+	if err != nil {
+		log.Errorf("unable to save service instance - %v", err)
+		return err
+	}
+	return nil
+}
+
+// DeleteServiceInstance - Delete the service instance for an service instance id.
+func (d *Dao) DeleteServiceInstance(id string) error {
+	log.Debugf("Dao::DeleteServiceInstance -> [ %s ]", id)
+	return d.client.ServiceInstances(d.namespace).Delete(id, &metav1.DeleteOptions{})
+}
+
+// GetBindInstance - Retrieve a specific bind instance from the kvp API
+func (d *Dao) GetBindInstance(id string) (*apb.BindInstance, error) {
+	log.Debugf("get binidng instance: %v", id)
+	bi, err := d.client.ServiceBindings(d.namespace).Get(id, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	return convertServiceBindingToAPB(bi.Spec, bi.GetName()), nil
+}
+
+// SetBindInstance - Set the bind instance for id in the kvp API.
+func (d *Dao) SetBindInstance(id string, bindInstance *apb.BindInstance) error {
+	log.Debugf("set binding instance: %v", id)
+	b := convertServiceBindingToCRD(bindInstance)
+	bi := v1.ServiceBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      id,
+			Namespace: d.namespace,
+		},
+		Spec: b,
+	}
+	_, err := d.client.ServiceBindings(d.namespace).Create(&bi)
+	if err != nil {
+		log.Errorf("unable to save service binding - %v", err)
+		return err
+	}
+	return nil
+}
+
+// DeleteBindInstance - Delete the binding instance for an id in the kvp API.
+func (d *Dao) DeleteBindInstance(id string) error {
+	log.Debugf("Dao::DeleteBindInstance -> [ %s ]", id)
+	err := d.client.ServiceBindings(d.namespace).Delete(id, &metav1.DeleteOptions{})
+	return err
+}
+
+// SetState - Set the Job State in the kvp API for id.
+func (d *Dao) SetState(instanceID string, state apb.JobState) (string, error) {
+	log.Debugf("set job state for instance: %v token: %v", instanceID, state.Token)
+	j := convertJobStateToCRD(&state)
+	if js, err := d.client.JobStates(d.namespace).Get(state.Token, metav1.GetOptions{}); err == nil {
+		js.Spec = j
+		js.ObjectMeta.Labels[jobStateLabel] = fmt.Sprintf("%v", convertStateToCRD(state.State))
```

Comment:
Why are we setting the labels here? Selector filtering? That's actually pretty slick.